### PR TITLE
fix: DateToDay

### PR DESCRIPTION
**Pull request by Chetan07j**
[Open in Gitpod](https://gitpod.io/from-referrer/)&nbsp;[know more](https://www.gitpod.io/docs/pull-requests/)

### Describe your change:

- [X] Fix a bug or typo in an existing algorithm?

### Checklist:

- [X] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Javascript/blob/master/CONTRIBUTING.md).
- [X] This pull request is all my own work -- I have not plagiarized.
- [X] I know that pull requests will not be merged if they fail the automated tests.
- [X] This PR only changes one algorithm file. To ease review, please open separate PRs for separate algorithms.
- [x] All new JavaScript files are placed inside an existing directory.
- [x] All filenames should use the UpperCamelCase (PascalCase) style. There should be no spaces in filenames.
      **Example:**`UserProfile.js` is allowed but `userprofile.js`,`Userprofile.js`,`user-Profile.js`,`userProfile.js` are not
- [X] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

**Comment by Chetan07j** [comment]
@appgurueu Could you please share what are your thoughts on this PR?
It will be really helpful for me if there are any improvements I can take care of next time.
Thanks

File: Conversions/DateToDay.js [hunk 1]
```diff
@@ -9,34 +9,34 @@
     algorithm shown below gives us the number of the day and
     finally converts it to the name of the day.
 
-    Algorithm & Explanation : https://cs.uwaterloo.ca/~alopez-o/math-faq/node73.html
+    Algorithm & Explanation : https://en.wikipedia.org/wiki/Zeller%27s_congruence
 */
 
 // March is taken as the first month of the year.
 const calcMonthList = {
```
Comment (appgurueu) [inline Conversions/DateToDay.js:16]: This mapping should (1) use an array and (2) definitely needs a comment; "March is taken as first month of the year" explains nothing. I prefer the PRs that get rid of it entirely.

File: Conversions/DateToDay.js [hunk 2]
```diff
@@ -9,34 +9,34 @@
...
+  6: 6,
+  7: 7,
+  8: 8,
+  9: 9,
+  10: 10,
+  11: 11,
+  12: 12
 }
 
 // show the week day in a number : Sunday - Saturday => 0 - 6
```
Comment (appgurueu) [inline Conversions/DateToDay.js:31]: This comment is now wrong.

File: Conversions/DateToDay.js [hunk 3]
```diff
@@ -45,18 +45,25 @@ const DateToDay = (date) => {
...
   // check the data are valid or not.
-  if (day < 0 || day > 31 || month > 12 || month < 0) {
+  if (day < 1 || day > 31 || month > 12 || month < 1) {
     return new TypeError('Date is not valid.')
   }
+
+  // In case of Jan and Feb, we consider it as previous year
+  // e.g., 1/1/1987 here year is 1986 (-1)
+  if (month < 3) {
+    year -= 1
```
Comment (appgurueu) [inline Conversions/DateToDay.js:57]: `year--`

File: Conversions/DateToDay.js [hunk 4]
```diff
@@ -9,34 +9,34 @@
...
+  7: 7,
+  8: 8,
+  9: 9,
+  10: 10,
+  11: 11,
+  12: 12
 }
 
 // show the week day in a number : Sunday - Saturday => 0 - 6
 const daysNameList = { // weeks-day
```
Comment (appgurueu) [inline Conversions/DateToDay.js:32]: This should be an array.

File: Conversions/DateToDay.js (hunk 1 above)
Comment (Chetan07j) [reply Conversions/DateToDay.js]: Noted.
Updated code as per comment.

I had thought of making it array as you suggested, like below:

```js
/**
 * This array holds months number
 * Considering March as the first month of the year
 * For January & February we store value as 13 & 14 respectively,
 * we consider these two from last year
 * And for other months we store exact month number
 */
const monthArray = [13, 14, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]
```

/# And then use it as **monthArray[month - 1]**
```js
const weekDay = Math.abs((day + Math.floor((monthArray[month - 1] + 1) * 2.6) + yearDigit + (yearDigit / 4) + (century / 4) - (2 * century)) % 7)
```

Then updated the code and remove this dependency completely.

File: Conversions/DateToDay.js (hunk 2 above)
Comment (Chetan07j) [reply Conversions/DateToDay.js]: Noted and updated comment to more meaningful

File: Conversions/DateToDay.js (hunk 3 above)
Comment (Chetan07j) [reply Conversions/DateToDay.js]: Done.

File: Conversions/DateToDay.js (hunk 4 above)
Comment (Chetan07j) [reply Conversions/DateToDay.js]: Done. Made it as an array, very good suggestion.

File: Conversions/DateToDay.js [hunk 5]
```diff
@@ -51,19 +27,25 @@ const DateToDay = (date) => {
...
+    month += 12
   }
 
   // divide year to century and yearDigit value.
   const yearDigit = (year % 100)
   const century = Math.floor(year / 100)
+
   // Apply the algorithm shown above
-  const weekDay = Math.abs((day + Math.floor((calcMonthList[month] + 1) * 2.6) + yearDigit + (yearDigit / 4) + (century / 4) - (2 * century)) % 7)
+  const weekDay = Math.abs((day + Math.floor((month + 1) * 2.6) + yearDigit + (yearDigit / 4) + (century / 4) - (2 * century)) % 7)
```
Comment (appgurueu) [inline Conversions/DateToDay.js:44]: Which formula are you implementing here?

File: Conversions/DateToDay.js [hunk 6]
```diff
@@ -51,19 +27,25 @@ const DateToDay = (date) => {
...
+
+  // There is possibility that weekDay is in decimal, so Math.floor() needed
   // return the weekDay name.
-  return daysNameList[Math.floor(weekDay)]
+  return daysNameArr[Math.floor(weekDay)]
```
Comment (appgurueu) [inline Conversions/DateToDay.js:48]: None of the formulas I've seen is `floor(abs(...))`.

File: Conversions/DateToDay.js (hunk 5 above)
Comment (Chetan07j) [reply Conversions/DateToDay.js:44]: https://en.wikipedia.org/wiki/Zeller%27s_congruence

Comment (appgurueu) [reply Conversions/DateToDay.js:44]: You didn't answer my question. Which formula - there are multiple formulas for different calendars on that page. You will want the one which (1) doesn't have problems with JS mod and (2) is for the Gregorian (not the Julian) calendar:
Formula

Comment (Chetan07j) [reply Conversions/DateToDay.js]: Sorry, misunderstood it.
I have implemented this one:
image

Comment (appgurueu) [reply Conversions/DateToDay.js]: No, you did not:

```javascript
Math.floor((day + Math.floor((month + 1) * 2.6) + yearDigit + (yearDigit / 4) + (century / 4) - (2 * century)) % 7)
```

* Why is there a `floor` around it? There is no floor around the formula.
* There must be a `floor` around `K/4` and `J/4`; you **may not** apply this afterwards
* This won't work for all cases due to JavaScript's [modulo operator](https://en.wikipedia.org/wiki/Zeller's_congruence#Basic_modification)

Comment (Chetan07j) [reply Conversions/DateToDay.js]: Updated, could you please check

Comment (appgurueu) [reply Conversions/DateToDay.js]: You have addressed the first two points, but how have you addressed the last point?

Comment (Chetan07j) [reply Conversions/DateToDay.js]: I have now addressed point 3 as well, as per explanation given [here](https://en.wikipedia.org/wiki/Zeller's_congruence#Basic_modification)

> The formulas rely on the mathematician's definition of [modulo](https://en.wikipedia.org/wiki/Modulo_operation) division, which means that −2 mod 7 is equal to positive 5. Unfortunately, in the truncating way most computer languages implement the remainder function, −2 mod 7 returns a result of −2. So, to implement Zeller's congruence on a computer, the formulas should be altered slightly to ensure a positive numerator. The simplest way to do this is to replace − 2J by + 5J

image

File: Conversions/DateToDay.js [hunk 7]
```diff
@@ -9,54 +9,43 @@
...
+  if (month < 3) {
+    year--
+    month += 12
+  }
+
   // divide year to century and yearDigit value.
   const yearDigit = (year % 100)
   const century = Math.floor(year / 100)
+
   // Apply the algorithm shown above
```
Comment (appgurueu) [inline Conversions/DateToDay.js:43]: Please add a note explaining that Java's modulo operator requires an adaption of the formula.

File: Conversions/DateToDay.js [hunk 8]
```diff
@@ -9,54 +9,43 @@
...
-  const weekDay = Math.abs((day + Math.floor((2.6 * calcMonthList[month]) - 0.2) - (2 * century) + yearDigit + Math.floor(yearDigit / 4) + Math.floor(century / 4)) % 7)
+  const weekDay = (day + Math.floor((month + 1) * 2.6) + yearDigit + Math.floor(yearDigit / 4) + Math.floor(century / 4) + (5 * century)) % 7
+
+  // There is possibility that weekDay is in decimal, so Math.floor() needed
```
Comment (appgurueu) [inline Conversions/DateToDay.js:46]: This comment is now outdated.

File: Conversions/DateToDay.js (hunk 7 above)
Comment (Chetan07j) [reply Conversions/DateToDay.js]: Done. Updated comment

File: Conversions/DateToDay.js (hunk 8 above)
Comment (Chetan07j) [reply Conversions/DateToDay.js]: Done.

File: Conversions/DateToDay.js [hunk 9]
```diff
@@ -40,10 +40,21 @@ const DateToDay = (date) => {
   const yearDigit = (year % 100)
   const century = Math.floor(year / 100)
 
-  // Apply the algorithm shown above
+  /**
```
Comment (appgurueu) [inline Conversions/DateToDay.js:43]: Don't copy verbatim from Wikipedia.

Comment (Chetan07j) [reply Conversions/DateToDay.js:43]: I have tried to explain the modulo division, please check and let me know if it makes sense.
Also, I have updated the commit message as `comment update` instead doc (_which was incorrect in this case_).

**Review by appgurueu — APPROVED** [review]

**Review by raklaptudirm — APPROVED** [review]

**Comment by Chetan07j** [comment]
Thank you @appgurueu @raklaptudirm for merging my PR.

Thank you @appgurueu for all your valuable review comments and suggestion, I really learned a lot from you.

**Pull request merged**